src/main/range-util.js

```diff
@@ -0,0 +1,218 @@
+"use strict";
+
+const comments = require("./comments");
+
+function findSiblingAncestors(startNodeAndParents, endNodeAndParents, opts) {
+  let resultStartNode = startNodeAndParents.node;
+  let resultEndNode = endNodeAndParents.node;
+
+  if (resultStartNode === resultEndNode) {
+    return {
+      startNode: resultStartNode,
+      endNode: resultEndNode
+    };
+  }
+
+  for (const endParent of endNodeAndParents.parentNodes) {
+    if (
+      endParent.type !== "Program" &&
+      endParent.type !== "File" &&
+      opts.locStart(endParent) >= opts.locStart(startNodeAndParents.node)
+    ) {
+      resultEndNode = endParent;
+    } else {
+      break;
+    }
+  }
+
+  for (const startParent of startNodeAndParents.parentNodes) {
+    if (
+      startParent.type !== "Program" &&
+      startParent.type !== "File" &&
+      opts.locEnd(startParent) <= opts.locEnd(endNodeAndParents.node)
+    ) {
+      resultStartNode = startParent;
+    } else {
+      break;
+    }
+  }
+
+  return {
+    startNode: resultStartNode,
+    endNode: resultEndNode
+  };
+}
+
+function findNodeAtOffset(node, offset, options, predicate, parentNodes) {
+  predicate = predicate || (() => true);
+  parentNodes = parentNodes || [];
+  const start = options.locStart(node, options.locStart);
+  const end = options.locEnd(node, options.locEnd);
+  if (start <= offset && offset <= end) {
+    for (const childNode of comments.getSortedChildNodes(
+      node,
+      undefined /* text */,
+      options
+    )) {
+      const childResult = findNodeAtOffset(
+        childNode,
+        offset,
+        options,
+        predicate,
+        [node].concat(parentNodes)
+      );
+      if (childResult) {
+        return childResult;
+      }
+    }
+
+    if (predicate(node)) {
+      return {
+        node: node,
+        parentNodes: parentNodes
+      };
+    }
+  }
+}
+
+// See https://www.ecma-international.org/ecma-262/5.1/#sec-A.5
+function isSourceElement(opts, node) {
```

Comment:
Should this be extracted to the plugins?

Reply:
I think so, but no plugin asked for this yet